keps/sig-api-machinery/xxx-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,764 @@
+
+In order to mitigate the above problem, we propose a couple different changes.
+
+[the documentation]: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list
+
+#### Reduce the number of requests during initialization
+
+As a first step, we will try to reduce the number of requests that are even reaching
+kube-apiserver with unitialized watchcache. To achieve it, we will introduce a new
+PostStartHook that will wait for watchcache of all *builtin* resources to be initialized.
+
+Why this should help with reducing the number of requests? The whole idea behind
+kube-apiserver readiness is that Kubernetes installations should be configured so that
+requests are not sent to it until kube-apiserver becomes ready. The above will allow us
+to delay this moment until watchcache is initialized.
+
+Why PostStartHook instead of /readyz check? Watchcache is per-type layer and we want
+to avoid marking whole kube-apiserver as not-ready if watchcache for one of resources
+requires reinitialization. We will handle those cases differently.
+
+Finally, we suggest starting just with builtin resources, as CRDs can be created at
+any time, making it harder to implement with unclear gain. If it appears to not be
+enough, this decision can easily be revisited in the future.
+
+#### Reject hanging watches
+
+To mitigate the problem of starving certain PriorityLevels, instead of having the
+watch hang and wait for watchcache initialization, we will simply reject it with
+`Too Many Requests` 429 http code.
+The reason for using this code is that it's already [properly handled] by our client
+libraries (e.g. reflector) and will not risk all of them suddenly falling back
+to list requests. Moreover, given that 429 is a long supported http code, we expect
+that other clients should also properly support it.
+
+[properly handled]: https://github.com/kubernetes/kubernetes/blob/41e706600aea7468f486150d951d3b8948ce89d5/staging/src/k8s.io/client-go/tools/cache/reflector.go#L910-L920
+
+#### Delegate get requests to etcd
+
+Since GET requests have bounded cost, we always estimate their weight the same
+no matter if they are served from etcd or from cache (the latency would be
+different in those cases though, so the actual seat-seconds cost too) and
+finally given we already did some work to process this request, we would
+simply delegate those requests to etcd.
+
+#### Adjust what lists are delegated to etcd
+
+It's tempting to reject all list requests with 429 the same way as watches.
+However, that would also have a negative consequences, e.g. by slowing down
+kube-apiserver initialization (kube-apiserver on startup need to initialize
+its own informers by loop-client, and by rejecting all lists until watchcache
+is initialized, we would effectively block that initialization).
+
+While we don't may decide to adjust the approach here, we will start with
+the approach that we will delegate to etcd the requests that (a) are not
```

Comment:
```suggestion
While we didn't decide to reject _all_ list requests with 429 here, we will start with
an approach where we only delegate to etcd the requests that (a) are not
```

(wording suggestion only.. please only accept if this matches what was intended)

Reply:
it matches - thanks for clarifying